src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php

```diff
@@ -125,7 +128,7 @@ public function preSubmit(FormEvent $event)
                 if (!$form->has($name)) {
                     $form->add($name, $this->type, array_replace(array(
                         'property_path' => '['.$name.']',
-                    ), $this->options));
+                    ), $this->getOptionsForEntry($value)));
```

Comment:
I'm going to add a test for it anyway, as for the behavior of `prototype_data` and `empty_data` on submit to avoid regressions such as #17992.

I keep investigating, I think this feature can be useful and doable even if it needs a more global refactoring of `CollectionType`.

It may not be finished for 3.1 though, but I'll try to word on that this week-end.